### PR TITLE
Add ability to delete CRD associated with App in Cluster Tools explorer

### PR DESCRIPTION
**Pull request by jordojordo**
rancher/dashboard#3811 
Apps remaining in the navigation after being deleted from the cluster tools explorer is caused by the remaining CustomResourceDefinition of the app that is not deleted. The CRD is created automatically when an app is installed.
In the Cluster Tools explorer I've added a checkbox in the prompt to remove the app which gives the ability to delete the CRD as well.

### CODE REVIEW

File: components/PromptRemoveApp.vue [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+          res += i === this.toRemove.length - 2 ? ' and ' : ', ';
+        }
+
+        return res;
+      }, '');
+    },
+
+    checkForCrd() {
+      if (this.allInstalled !== null) {
+        const crd = this.allInstalled.find(installed => installed.metadata.name === `${ this.names }-crd`);
```
Comment (vincent99) [inline components/PromptRemoveApp.vue:157]: We should set an annotation on the app that says what other things were auto-installed for it.  Assuming a naming-scheme exists is fragile, and not always true (there are charts which auto-install things other than CRDs).

**Review by richard-cox — CHANGES_REQUESTED** [review]
- There was an old rather hardcoded way of showing modals (promptRemove) and a newer way that removes a lot of the plumbing (promptModal). I'd recommend switching over to the newer way, this removes changes to store/action-menu.js, plugins/steve/actions.js and PromptRemoveApp changes in pages/c/_cluster/explorer/tools/index.vue. If you search for where the action `promptModal` is dispatched you should find some good examples of how it works.
- PromptRemove vs PromptRemoveApp. There's a LOT of duplicated code. You could use a `mixin` that both use. That doesn't remove the duplicated template but would be an improvement. The only other option i can think of is if PromptRemove became more generic (exposing slots/methods that differ between the two) and used as a component in PromptRemoveApp, though that gets rather noodley. Might be worth raising in the next sync up, someone might know a better pattern.

File: components/PromptRemoveApp.vue (hunk 1 above)
Comment (richard-cox) [reply components/PromptRemoveApp.vue:157]: Yeah this is something that needs more polish @jordojordo bring this up in the next sync up and we can find a path forward

**Review by richard-cox — CHANGES_REQUESTED** [review]
Remember, as we don't squash on merge to master your PR's commit history will make it in. So to avoid 'merge branch x into y' commits make sure to use `rebase` instead. For example `git rebase <upstream>/<branch>`.
For this PR, to remove those, could you squash them into one by using something like `git rebase -i <upstream>/<master>`. Just ping me on slack if you'd like to run through that flow together.
After that's gone in I'll give it a quick test, otherwise code looks good. It would have been nice to not include the app specific code in the generic promptRemove, but as discussed and agreed together there's not much code and the alternatives are expensive.

**Comment by nwmac** [comment]
@jordojordo Can you take another look at this

**Comment by jordojordo** [comment]
@richard-cox I've cleaned up what I had written before and fixed the issue where the CRD wasn't being selected properly. Now when if you delete both the main resource and the associated CRD resource (tested with Longhorn), the menu item will be removed.

File: components/PromptRemove.vue [hunk 2]
```diff
@@ -145,6 +154,9 @@ export default {
   watch:    {
     showPromptRemove(show) {
       if (show) {
+        if (this.currentRouter?.currentRoute?.name === 'c-cluster-explorer-tools' && this.toRemove[0].type === CATALOG.APP && this.toRemove[0].spec?.chart?.metadata?.annotations[CATALOG_ANNOTATIONS.AUTO_INSTALL]) {
```
Comment (richard-cox) [inline components/PromptRemove.vue:157]: Just a styling comment, this is a pretty long line and should be split up, would suggest each clause is on a new one

File: components/PromptRemove.vue [hunk 3]
```diff
@@ -2,23 +2,32 @@
...
+      randomPosition:      Math.random(),
+      confirmName:         '',
+      error:               '',
+      warning:             '',
+      preventDelete:       false,
+      removeComponent:     this.$store.getters['type-map/importCustomPromptRemove'](resource),
+      deleteCrd:           false,
+      allInstalled:        null,
+      chartsToRemoveIsApp: false,
+      chartsDeleteCrd:     false
```
Comment (richard-cox) [inline components/PromptRemove.vue:30]: This is set/reset but isn't used anywhere

File: plugins/steve/resource-instance.js [hunk 4]
```diff
@@ -0,0 +1,1645 @@
+import compact from 'lodash/compact';
```
Comment (richard-cox) [inline plugins/steve/resource-instance.js:1]: `resource-instance` now doesn't exist, you'll need to port changes to `resource-class`. Beat in mind also anything Steve/Norman specific may be better placed in

**Review by richard-cox — CHANGES_REQUESTED** [review]
Haven't run this yet, but there were a few odd things in code, mainly properties that don't seem to be used / plumbed in

File: components/PromptRemove.vue (hunk 3 above)
Comment (jordojordo) [reply components/PromptRemove.vue:28]: When I wrote this the first time I was using two different variables to achieve the same thing. So I removed `deleteCrd` and replaced where I was arguing against it with `chartsDeleteCrd`. Now if the checkbox is unchecked, `chartsDeleteCrd` will be false and the selected CRD will be removed from the `toRemove` array.

File: components/PromptRemove.vue [hunk 5]
```diff
@@ -2,23 +2,32 @@
...
-      removeComponent: this.$store.getters['type-map/importCustomPromptRemove'](resource)
+      hasCustomRemove:     false,
+      randomPosition:      Math.random(),
+      confirmName:         '',
+      error:               '',
+      warning:             '',
+      preventDelete:       false,
+      removeComponent:     this.$store.getters['type-map/importCustomPromptRemove'](resource),
+      deleteCrd:           false,
+      allInstalled:        null,
```
[the remark this answers is not in the document]
Comment (jordojordo) [reply components/PromptRemove.vue]: It wasn't used anywhere and has been removed.

File: components/PromptRemove.vue [hunk 6]
```diff
@@ -2,23 +2,32 @@
...
-      preventDelete:   false,
-      removeComponent: this.$store.getters['type-map/importCustomPromptRemove'](resource)
+      hasCustomRemove:     false,
+      randomPosition:      Math.random(),
+      confirmName:         '',
+      error:               '',
+      warning:             '',
+      preventDelete:       false,
+      removeComponent:     this.$store.getters['type-map/importCustomPromptRemove'](resource),
+      deleteCrd:           false,
```
[the remark this answers is not in the document]
Comment (jordojordo) [reply components/PromptRemove.vue]: `deleteCrd` has been removed.

File: components/PromptRemove.vue (hunk 2 above)
Comment (jordojordo) [reply components/PromptRemove.vue]: I cleaned this up a bit.

File: plugins/steve/resource-instance.js (hunk 4 above)
Comment (jordojordo) [reply plugins/steve/resource-instance.js]: Fixed :)

**Review by richard-cox — APPROVED** [review]
I noticed that the app still remains in the side nav even after deleting both the core and `<x>-crd` app (the page needs refreshing for them to be removed). This happened when directly deleting them as well from the charts page so I've created a separate issue - https://github.com/rancher/dashboard/issues/4849

**Pull request merged**